api/src/main/java/org/apache/iceberg/types/Types.java

```diff
@@ -414,42 +415,96 @@ public int hashCode() {
+      return new NestedField(true, id, name, type, defaultValue, null);
     }
 
     public static NestedField required(int id, String name, Type type) {
-      return new NestedField(false, id, name, type, null);
+      return new NestedField(false, id, name, type, null, null);
     }
 
     public static NestedField required(int id, String name, Type type, String doc) {
-      return new NestedField(false, id, name, type, doc);
+      return new NestedField(false, id, name, type, null, doc);
+    }
+
+    public static NestedField required(int id, String name, Type type, Object defaultValue, String doc) {
+      validateDefaultValueForRequiredField(defaultValue, name);
+      return new NestedField(false, id, name, type, defaultValue, doc);
+    }
+
+    public static NestedField requiredWithDefault(int id, String name, Type type, Object defaultValue) {
+      validateDefaultValueForRequiredField(defaultValue, name);
+      return new NestedField(false, id, name, type, defaultValue, null);
     }
 
     public static NestedField of(int id, boolean isOptional, String name, Type type) {
-      return new NestedField(isOptional, id, name, type, null);
+      return new NestedField(isOptional, id, name, type, null, null);
     }
 
     public static NestedField of(int id, boolean isOptional, String name, Type type, String doc) {
-      return new NestedField(isOptional, id, name, type, doc);
+      return new NestedField(isOptional, id, name, type, null, doc);
+    }
+
+    public static NestedField of(int id, boolean isOptional, String name, Type type, Object defaultValue, String doc) {
+      return new NestedField(isOptional, id, name, type, defaultValue, doc);
+    }
+
+    private static void validateDefaultValueForRequiredField(Object defaultValue, String fieldName) {
+      Preconditions.checkArgument(defaultValue != null,
+          "Cannot create NestedField with a null default for a required field " +  fieldName);
+    }
+
+    private static void validateDefaultValue(Object defaultValue, Type type) {
+      if (defaultValue == null) {
+        return;
+      }
+      if (type.isStructType()) {
+        Preconditions.checkArgument(List.class.isInstance(defaultValue),
+            "defaultValue should be a List of Objects for StructType");
+        StructType.validateDefaultValue((List) defaultValue, type.asStructType().fieldList);
+      } else if (type.isListType()) {
+        Preconditions.checkArgument(defaultValue.getClass().isArray(),
+            "defaultValue should be an array of Objects for ListType");
+        ListType.validateDefaultValue((Object[]) defaultValue, type.asListType().elementField);
+      } else if (type.isMapType()) {
+        Preconditions.checkArgument(Map.class.isInstance(defaultValue),
+            "defaultValue should be an instance of Map for MapType");
+        MapType.validateDefaultValue((Map<Object, Object>) defaultValue,
```

Comment:
Can we eliminate `XType.validateDefaultValue` and have this method call itself recursively?

Reply:
done